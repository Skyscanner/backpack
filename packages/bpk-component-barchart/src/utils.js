/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2020 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const center = scale => {
  let offset = scale.bandwidth() / 2;
  if (scale.round()) {
    offset = Math.round(offset);
  }
  return d => scale(d) + offset;
};

const identity = x => x;

const remToPx = value => {
  let parsed = null;

  if (/rem$/.test(value)) {
    parsed = parseFloat(value.replace(/rem/, '')) * 16;
  }

  return parsed || null;
};

export { center, identity, remToPx };
