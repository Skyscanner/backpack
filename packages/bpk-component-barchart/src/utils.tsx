/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

// Using Function here as scale refers to the d3-scale 3rd party library
const center = (scale: Function) => {
  let offset = scale.bandwidth() / 2;
  if (scale.round()) {
    offset = Math.round(offset);
  }
  return (d: number) => scale(d) + offset;
};

// Using any here as x can be any form see utils-test.js
const identity = (x: any) => x;

const remToPx = (value: string) => {
  let parsed = 0;

  if (/rem$/.test(value)) {
    parsed = parseFloat(value.replace(/rem/, '')) * 16;
  }

  return parsed;
};

export { center, identity, remToPx };
