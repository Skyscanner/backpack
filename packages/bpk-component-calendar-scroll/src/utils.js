/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  calendarDaySize,
  calendarDaySpacing,
} from 'bpk-tokens/tokens/base.es6';

const remToPx = value => {
  let parsed = null;

  if (/rem$/.test(value)) {
    parsed = parseFloat(value.replace(/rem/, '')) * 16;
  }

  return parsed || null;
};

export const getCalendarGridWidth = () =>
  7 * (remToPx(calendarDaySize) + remToPx(calendarDaySpacing));

export const getTransformStyles = transformValue => {
  const transform = `translateX(${transformValue}px)`;
  return {
    transform,
    msTransform: transform,
    MozTransform: transform,
    WebkitTransform: transform,
  };
};

export const getScriptDirection = () => {
  const html = document.querySelector('html');
  return window.getComputedStyle(html, null).getPropertyValue('direction');
};

export const isTransitionEndSupported = () =>
  !!(typeof window !== 'undefined' && 'TransitionEvent' in window);
