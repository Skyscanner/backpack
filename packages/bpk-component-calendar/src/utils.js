/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  calendarDaySize,
  calendarDaySpacing,
} from '@skyscanner/bpk-foundations-web/tokens/base.es6';

const CSS_UNIT_REGEX = /(^[+-]?(?:\d*\.)?\d+)(.+)/i;

const splitToken = (value) => {
  const match = value.match(CSS_UNIT_REGEX);
  if (!match) {
    throw new Error(
      `Invalid token value. Expecting a valid css unit, got ${value}`,
    );
  }
  const [_, val, unit] = match; // eslint-disable-line no-unused-vars
  return [parseFloat(val), unit];
};

export const getCalendarGridWidth = (multiplier = 1) => {
  const [sizeValue, sizeUnit] = splitToken(calendarDaySize);
  const [spacingValue, spacingUnit] = splitToken(calendarDaySpacing);

  if (sizeUnit !== spacingUnit) {
    throw new Error(
      `'calendarDaySize' and 'calendarDaySpacing' must use the same unit. Got ${sizeUnit} and ${spacingUnit}`,
    );
  }
  const width = multiplier * (7 * (sizeValue + 2 * spacingValue));
  return `${width}${sizeUnit}`;
};

export const getTransformStyles = (transformValue) => {
  const transform = `translateX(${transformValue})`;
  return {
    transform,
    msTransform: transform,
    MozTransform: transform,
    WebkitTransform: transform,
  };
};

export const isTransitionEndSupported = () =>
  !!(typeof window !== 'undefined' && 'TransitionEvent' in window);
