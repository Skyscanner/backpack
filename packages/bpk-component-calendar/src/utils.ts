/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  calendarDaySize,
  calendarDaySpacing,
} from '@skyscanner/bpk-foundations-web/tokens/base.es6';

const CSS_UNIT_REGEX = /(^[+-]?(?:\d*\.)?\d+)(.+)/i;

const splitToken = (value: string) => {
  const match = value.match(CSS_UNIT_REGEX);
  if (!match) {
    throw new Error(
      `Invalid token value. Expecting a valid css unit, got ${value}`,
    );
  }
  const [_, val, unit] = match;
  return [parseFloat(val), unit];
};

export const getCalendarGridWidth = (multiplier = 1) => {
  const [sizeValue, sizeUnit] = splitToken(calendarDaySize);
  const [spacingValue, spacingUnit] = splitToken(calendarDaySpacing);

  if (sizeUnit !== spacingUnit) {
    throw new Error(
      `'calendarDaySize' and 'calendarDaySpacing' must use the same unit. Got ${sizeUnit} and ${spacingUnit}`,
    );
  }
  // @ts-expect-error TS doesn't correctly read the size and spacing values type i.e. numbers as returned by parseFloat
  const width = multiplier * (7 * (sizeValue + 2 * spacingValue));
  return `${width}${sizeUnit}`;
};

export const getTransformStyles = (transformValue: string) => {
  const transform = `translateX(${transformValue})`;
  return {
    transform,
    msTransform: transform,
    MozTransform: transform,
    WebkitTransform: transform,
  };
};

export const isTransitionEndSupported = () =>
  !!(typeof window !== 'undefined' && 'TransitionEvent' in window);

/**
 * Memoizes a given function to optimize performance by caching its results based on input arguments.
 *
 * @template T - The type of the function to be memoized. Must be a function.
 * @param {T} fn - The function to be memoized.
 * @returns {T} - The memoized version of the input function. Subsequent calls with the same arguments
 *                will return the cached result, avoiding redundant computations.
 *
 * @example
 * // Simple usage:
 * const add = (a: number, b: number) => a + b;
 * const memoizedAdd = memoize(add);
 *
 * console.log(memoizedAdd(1, 2)); // Computes and caches result: 3
 * console.log(memoizedAdd(1, 2)); // Fetches from cache: 3
 *
 * @example
 * // With a more complex function:
 * const slowFunction = (num: number) => {
 *   console.log('Computing...');
 *   return num * 2;
 * };
 * const memoizedSlowFunction = memoize(slowFunction);
 *
 * console.log(memoizedSlowFunction(5)); // Logs "Computing...", returns 10
 * console.log(memoizedSlowFunction(5)); // Fetches from cache: 10
 *
 * @remarks
 * - This implementation uses a `Map` object and `JSON.stringify` to create cache keys.
 * - Functions with non-primitive or cyclic arguments may not work as expected due to
 *   `JSON.stringify` limitations.
 * - The cache is stored in memory and will grow indefinitely unless manually managed or
 *   the function goes out of scope.
 */
export function memoize<T extends (...args: any[]) => any>(fn: T): T {
  const cache = new Map<string, ReturnType<T>>();

  function memoizedFunction(...args: Parameters<T>): ReturnType<T> {
    const key = JSON.stringify(args);
    if (cache.has(key)) {
      return cache.get(key) as ReturnType<T>;
    }
    const result = fn(...args);
    cache.set(key, result);
    return result;
  }

  return memoizedFunction as T;
}
