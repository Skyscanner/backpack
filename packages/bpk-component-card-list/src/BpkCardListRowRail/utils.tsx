/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2022 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useEffect } from 'react';

export function setA11yTabIndex(
  el: HTMLDivElement | null,
  index: number,
  visibleRatios: number[],
) {
  console.log('visibleRatios', visibleRatios);
  if (!el) return;
  const focusableElements = el.querySelectorAll<HTMLElement>(
    'a, button, input, textarea, select, [tabindex]:not([tabindex="-1"])',
  );

  focusableElements.forEach((element: HTMLElement) => {
    const targetElement = element;
    targetElement.tabIndex = visibleRatios[index] >= 0.8 ? 0 : -1;
  });
}

export function useUpdateCurrentIndexByVisibility(
  visibleRatios: number[],
  setCurrentIndex: (index: number) => void,
  setStateTimeoutRef: React.MutableRefObject<NodeJS.Timeout | null>,
) {
  useEffect(() => {
    if (!visibleRatios || visibleRatios.length === 0) return;

    if (setStateTimeoutRef.current) clearTimeout(setStateTimeoutRef.current);

    const firstVisibleIndex = visibleRatios.findIndex((ratio) => ratio > 0);

    // when the first visible card is not fully visible
    // the scrolling hasn't ended
    // we don't set the current index

    if (visibleRatios[firstVisibleIndex] > 0.95) {
      // eslint-disable-next-line no-param-reassign
      setStateTimeoutRef.current = setTimeout(() => {
        setCurrentIndex(firstVisibleIndex);
      }, 150);
    }
  }, [visibleRatios]);
}

export function useScrollToCard(
  currentIndex: number,
  container: HTMLElement | null,
  cardRefs: React.MutableRefObject<Array<HTMLDivElement | null>>,
  stateScrollingLockRef: React.MutableRefObject<boolean>,
) {
  useEffect(() => {
    const isVisible =
      container &&
      container.getBoundingClientRect().bottom > 0 &&
      container.getBoundingClientRect().bottom <= window.innerHeight;
    if (!isVisible) return; // Escape from scrollIntoView if the container is not visible

    if (stateScrollingLockRef.current && stateScrollingLockRef.current === true)
      return;

    const targetCard = cardRefs.current[currentIndex];
    if (targetCard) {
      targetCard.scrollIntoView({
        behavior: 'smooth',
        block: 'nearest',
        inline: 'start',
      });
    }
  }, [currentIndex]);
}

export function useIntersectionObserver(
  { root, threshold }: IntersectionObserverInit,
  visibleRatios: number[],
  setVisibleRatios: React.Dispatch<React.SetStateAction<number[]>>,
) {
  let observer: IntersectionObserver | null = null;

  if (root) {
    observer = new IntersectionObserver(
      (entries) => {
        entries.forEach((entry) => {
          const index = Number(entry.target.getAttribute('data-index'));

          if (entry.intersectionRatio !== visibleRatios[index]) {
            setVisibleRatios((prev) => {
              const newVisibleRatios = [...prev];
              newVisibleRatios[index] = entry.intersectionRatio;
              return newVisibleRatios;
            });
          }
        });
      },
      { root, threshold },
    );
  }

  const observe = (element: HTMLElement | null, index: number) => {
    if (element && observer) {
      element.setAttribute('data-index', index.toString());
      observer.observe(element);
    }
  };

  return observe;
}
