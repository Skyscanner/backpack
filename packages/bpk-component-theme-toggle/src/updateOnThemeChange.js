/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { wrapDisplayName } from '../../bpk-react-utils';

import { getHtmlElement, THEME_CHANGE_EVENT } from './utils';

const updateOnThemeChange = (EnhancedComponent) => {
  class UpdateOnThemeChange extends Component {
    constructor() {
      super();
      this.state = {
        theme: null,
      };
    }

    componentDidMount() {
      getHtmlElement().addEventListener(
        THEME_CHANGE_EVENT,
        this.onThemeChange,
        false,
      );
    }

    componentWillUnmount() {
      getHtmlElement().removeEventListener(
        THEME_CHANGE_EVENT,
        this.onThemeChange,
        false,
      );
    }

    onThemeChange = (e) => {
      const { theme } = e.detail;
      this.setState({ theme });
      this.forceUpdate();
    };

    render() {
      return <EnhancedComponent theme={this.state.theme} {...this.props} />;
    }
  }

  UpdateOnThemeChange.displayName = wrapDisplayName(
    EnhancedComponent,
    'updateOnThemeChange',
  );

  UpdateOnThemeChange.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return UpdateOnThemeChange;
};

export default updateOnThemeChange;
