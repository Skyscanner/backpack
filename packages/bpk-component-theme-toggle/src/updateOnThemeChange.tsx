/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
import { Component } from 'react';

import { wrapDisplayName } from '../../bpk-react-utils';

import { getHtmlElement, THEME_CHANGE_EVENT } from './utils';

const updateOnThemeChange = (EnhancedComponent: any) => {
  class UpdateOnThemeChange extends Component {
    constructor() {
      // @ts-expect-error TS(2554) FIXME: Expected 1-2 arguments, but got 0.
      super();
      this.state = {
        theme: null,
      };
    }

    componentDidMount() {
      // @ts-expect-error TS(2531) FIXME: Object is possibly 'null'.
      getHtmlElement().addEventListener(
        THEME_CHANGE_EVENT,
        this.onThemeChange,
        false,
      );
    }

    componentWillUnmount() {
      // @ts-expect-error TS(2531) FIXME: Object is possibly 'null'.
      getHtmlElement().removeEventListener(
        THEME_CHANGE_EVENT,
        this.onThemeChange,
        false,
      );
    }

    onThemeChange = (e: any) => {
      const { theme } = e.detail;
      this.setState({ theme });
      this.forceUpdate();
    };

    render() {
      // @ts-expect-error TS(2339) FIXME: Property 'theme' does not exist on type 'Readonly<... Remove this comment to see the full error message
      return <EnhancedComponent theme={this.state.theme} {...this.props} />;
    }
  }

  // @ts-expect-error TS(2339) FIXME: Property 'displayName' does not exist on type 'typ... Remove this comment to see the full error message
  UpdateOnThemeChange.displayName = wrapDisplayName(
    EnhancedComponent,
    'updateOnThemeChange',
  );

  // @ts-expect-error TS(2339) FIXME: Property 'propTypes' does not exist on type 'typeo... Remove this comment to see the full error message
  UpdateOnThemeChange.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return UpdateOnThemeChange;
};

export default updateOnThemeChange;
