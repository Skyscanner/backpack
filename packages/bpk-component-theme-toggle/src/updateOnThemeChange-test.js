/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
import { render, fireEvent } from '@testing-library/react';

import updateOnThemeChange from './updateOnThemeChange';
import { THEME_CHANGE_EVENT, getHtmlElement } from './utils';

const Dummy = ({ children }) => <div>{children}</div>;

const EnhancedComponent = updateOnThemeChange(Dummy);

describe('EnhancedComponent', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <EnhancedComponent>
        <p>Children</p>
      </EnhancedComponent>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should force an update when receiving a theme change event', () => {
    render(
      <EnhancedComponent>
        <p />
      </EnhancedComponent>,
    );

    const forceUpdateSpy = jest.spyOn(
      EnhancedComponent.prototype,
      'forceUpdate',
    );
    expect(forceUpdateSpy).not.toHaveBeenCalled();

    fireEvent(
      getHtmlElement(),
      new CustomEvent(THEME_CHANGE_EVENT, { detail: { theme: null } }),
    );

    expect(forceUpdateSpy).toHaveBeenCalled();
  });
});

Dummy.propTypes = {
  children: PropTypes.node.isRequired,
};
