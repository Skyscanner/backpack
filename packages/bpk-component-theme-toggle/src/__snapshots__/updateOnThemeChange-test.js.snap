// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`EnhancedComponent should render correctly 1`] = `
<DocumentFragment>
  <div>
    <p>
      Children
    </p>
  </div>
</DocumentFragment>
`;
